### DISCUSSION TITLE
extract text pdf with text split into 2 columns with a table on the bottom

### DISCUSSION THREAD

**Discussion by maayansharon10** [jsvine/pdfplumber#751] [Ask for help with specific PDFs]
Hello, 
Thank you for a great library.
I've been trying to find a general way to extract text only - without tables, header or footer from random pdfs
My problem is that I cannot seem to detect whether a file has 2 columns text or when it's a table.
for example - I want to extract only the "main text" meaning without header, footer and the table below, so far with little success.
Would appreciate some help.
is it possible to code something that could extract this type of text for multiple different pdfs with pdfplumber? 
[AspenIRP_Final_November2021_page18.pdf](https://github.com/jsvine/pdfplumber/files/9896246/AspenIRP_Final_November2021_page18.pdf)

currently I've got this - 
`
import pdfplumber
pdf = pdfplumber.open("/content/AspenIRP_Final_November2021_page18.pdf")
p = pdf.pages[0] # Load the page.

tables = p.find_tables()
header_row = tables[0].rows[0].cells
vertical_lines = [cell[0] for cell in header_row] + [header_row[-1][2]]

ts = {
    "vertical_strategy": "explicit",
    "horizontal_strategy": "lines",
    "explicit_vertical_lines": vertical_lines,
}

bboxes = [table.bbox for table in p.find_tables(table_settings=ts)] # Get the bounding boxes of the tables on the page.

def not_within_bboxes(obj):
    """Check if the object is in any of the table's bbox."""
    def obj_in_bbox(_bbox):
        """See https://github.com/jsvine/pdfplumber/blob/stable/pdfplumber/table.py#L404"""
        v_mid = (obj["top"] + obj["bottom"]) / 2
        h_mid = (obj["x0"] + obj["x1"]) / 2
        x0, top, x1, bottom = _bbox
        return (h_mid >= x0) and (h_mid < x1) and (v_mid >= top) and (v_mid < bottom)
    return not any(obj_in_bbox(__bbox) for __bbox in bboxes)

print("Text outside the tables:")
print(p.filter(not_within_bboxes).extract_text())`

and result was :

`Text outside the tables:
 
 
ES-6   NOVEMBER 2021   FINAL   
| |`

# Extract text without graph
similar yet a bit different problem was when I was trying to extract only the "main text" from this page and could not ignore the graph since it is not a picture. is there a way to do so? 
[AspenIRP_Final_November2021_page15.pdf](https://github.com/jsvine/pdfplumber/files/9896289/AspenIRP_Final_November2021_page15.pdf)

Thank you

ps. Saw those discussions which were helpful but still didn't get to the exact solution https://github.com/jsvine/pdfplumber/issues/242 https://github.com/jsvine/pdfplumber/discussions/437

**Comment by jsvine** [comment]
Hi @maayansharon10, and thanks for your kind words about this library. If you're working with a consistent set of PDFs (e.g., all produced by the same software), there may be a deterministic approach to detecting column layouts. Otherwise, you're at the mercy of all the strange ways that people (and software packages) decide to construct PDFs. There might be probabilistic / ML approaches that could detect such layouts, but I'm not deeply familiar with them.
